# Paso 1 Cargar el Data Set

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Cargar el dataset
data = pd.read_csv('train.csv')


c:\Users\mile1\anaconda3\lib\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.0' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
c:\Users\mile1\anaconda3\lib\site-packages\pandas\core\arrays\masked.py:62: UserWarning: Pandas requires version '1.3.4' or newer of 'bottleneck' (version '1.3.2' currently installed).
  from pandas.core import (
c:\Users\mile1\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# Paso 2 Dividir el Data Set en Training y Test

In [2]:
# Dividir el dataset en características (X) y etiquetas (y)
X = data.iloc[:, 1:]
y = data.iloc[:, 0]

# Dividir el dataset en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Paso 3. Preprocesamiento de Datos

In [3]:
import numpy as np
from sklearn.preprocessing import StandardScaler, OneHotEncoder

# Normalizar los valores de píxeles
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [6]:
# Convertir y_train a un arreglo de numpy
y_train_array = np.array(y_train)
y_test_array= np.array(y_test)

In [7]:
# Transformar las etiquetas en formato one-hot encoding
encoder = OneHotEncoder()
y_train_encoded = encoder.fit_transform(y_train_array.reshape(-1, 1)).toarray()
y_test_encoded = encoder.transform(y_test_array.reshape(-1, 1)).toarray()

# Paso 4. Creación de la red neuronal artificial

In [8]:
import tensorflow as tf
from tensorflow import keras

# Definir la estructura de la red neuronal
model = keras.Sequential([
    keras.layers.Dense(128, activation='relu', input_shape=(784,)),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

# Compilar la red neuronal
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Paso 5. Entrenamiento de la red neuronal artificial

In [9]:
import tensorflow as tf
from tensorflow import keras

# Definir la estructura de la red neuronal
model = keras.Sequential([
    keras.layers.Dense(128, activation='relu', input_shape=(784,)),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

# Compilar la red neuronal
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Paso 6. Evaluación de la red neuronal artificial

In [14]:
# Obtener las predicciones de la red neuronal para el conjunto de prueba
y_test_pred = model.predict(X_test)

263/263 [==============================] - 1s 2ms/step


In [15]:
# Convertir las predicciones a etiquetas numéricas
y_test_pred_labels = np.argmax(y_test_pred, axis=1)

In [16]:
# Calcular las métricas de clasificación
accuracy = accuracy_score(y_test, y_test_pred_labels)
precision = precision_score(y_test, y_test_pred_labels, average='macro')
recall = recall_score(y_test, y_test_pred_labels, average='macro')
f1 = f1_score(y_test, y_test_pred_labels, average='macro')

c:\Users\mile1\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [17]:
# Mostrar las métricas de clasificación
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


Accuracy: 0.06904761904761905
Precision: 0.03916071347690282
Recall: 0.0747279332291608
F1 Score: 0.04563305889714333


# Paso 7. Red neuronal Convolucional

In [18]:
import tensorflow as tf
from tensorflow import keras

# Definir la estructura de la red neuronal convolucional
model = keras.Sequential([
    keras.layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)),
    keras.layers.MaxPooling2D(pool_size=(2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

# Compilar la red neuronal convolucional
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Paso 8. Entrenamiento de la red convolucional

In [22]:
import numpy as np

# Convert DataFrame to NumPy array
X_train_array = X_train.values
X_test_array = X_test.values

# Reshape the input data for CNN
X_train_reshaped = X_train_array.reshape(-1, 28, 28, 1)
X_test_reshaped = X_test_array.reshape(-1, 28, 28, 1)


In [26]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

# Convert DataFrame to NumPy array
X_train_array = X_train.values
X_test_array = X_test.values

# Reshape the input data for CNN
X_train_reshaped = X_train_array.reshape(-1, 28, 28, 1)
X_test_reshaped = X_test_array.reshape(-1, 28, 28, 1)

# Clear the TensorFlow graph
tf.keras.backend.clear_session()

# Define the structure of the convolutional neural network
model = keras.Sequential([
    keras.layers.Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)),
    keras.layers.MaxPooling2D(pool_size=(2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

# Compile the convolutional neural network
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(X_train_reshaped, y_train_encoded, batch_size=64, epochs=10, validation_data=(X_test_reshaped, y_test_encoded))


Epoch 1/10
525/525 [==============================] - 17s 30ms/step - loss: 1.1631 - accuracy: 0.9122 - val_loss: 0.1565 - val_accuracy: 0.9586
Epoch 2/10
525/525 [==============================] - 15s 29ms/step - loss: 0.0999 - accuracy: 0.9713 - val_loss: 0.1193 - val_accuracy: 0.9680
Epoch 3/10
525/525 [==============================] - 14s 28ms/step - loss: 0.0505 - accuracy: 0.9844 - val_loss: 0.1213 - val_accuracy: 0.9713
Epoch 4/10
525/525 [==============================] - 15s 28ms/step - loss: 0.0383 - accuracy: 0.9882 - val_loss: 0.1434 - val_accuracy: 0.9679
Epoch 5/10
525/525 [==============================] - 16s 31ms/step - loss: 0.0287 - accuracy: 0.9907 - val_loss: 0.1255 - val_accuracy: 0.9699
Epoch 6/10
525/525 [==============================] - 15s 28ms/step - loss: 0.0189 - accuracy: 0.9940 - val_loss: 0.1394 - val_accuracy: 0.9700
Epoch 7/10
525/525 [==============================] - 14s 27ms/step - loss: 0.0405 - accuracy: 0.9885 - val_loss: 0.1309 - val_accuracy:

# Paso 9. Evaluación del modelo convolucional

In [27]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Obtener las predicciones de la red neuronal convolucional para el conjunto de prueba
y_test_pred = model.predict(X_test_reshaped)

# Convertir las predicciones a etiquetas numéricas
y_test_pred_labels = np.argmax(y_test_pred, axis=1)

# Calcular las métricas de clasificación
accuracy = accuracy_score(y_test, y_test_pred_labels)
precision = precision_score(y_test, y_test_pred_labels, average='macro')
recall = recall_score(y_test, y_test_pred_labels, average='macro')
f1 = f1_score(y_test, y_test_pred_labels, average='macro')

# Mostrar las métricas de clasificación
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


263/263 [==============================] - 1s 4ms/step
Accuracy: 0.9742857142857143
Precision: 0.9748073000369999
Recall: 0.9738955656584759
F1 Score: 0.9742588670484628


# Paso 10. Comparación

Se concluye que el modelo más efectivo es la red neuronal Convolucional, debido al alto de accuracy que nos brinda (0.9742), mientas que la red neuronal artificial es completamente ineficiente con un accuracy de 0.069 para el manejo y clasificación de pixeles